File: src/main/java/ru/ensemplix/command/CommandDispatcher.java
package ru.ensemplix.command;

import com.google.common.base.CharMatcher;
import ru.ensemplix.command.argument.Argument;
import ru.ensemplix.command.argument.ArgumentParser;
import ru.ensemplix.command.exception.CommandAccessException;
import ru.ensemplix.command.exception.CommandException;
import ru.ensemplix.command.exception.CommandNotFoundException;

import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.lang.reflect.ParameterizedType;
import java.util.*;
import java.util.stream.Collectors;

import static com.google.common.base.Preconditions.checkNotNull;
import static ru.ensemplix.command.argument.ArgumentParser.*;

/**
 * Основной класс для работы с командами.
 */
public class CommandDispatcher {

    /**
     * Список автоматических дополнений команды.
     */
    protected static final Map<Class, CommandCompleter> completers = new HashMap<>();

    /**
     * Список команд.
     */
    protected static final Map<String, CommandHandler> commands = new HashMap<>();

    /**
     * Список парсеров в объекты.
     */
    protected static final Map<Class, ArgumentParser> parsers = new HashMap<>();

    /**
     * Нужно ли убирать первый символ("/", "!", "@") при выполнении команды.
     * По умолчанию символ убирается.
     */
    private final boolean removeFirstChar;

    public CommandDispatcher() {
        this(true);
    }

    public CommandDispatcher(boolean removeFirstChar) {
        this.removeFirstChar = removeFirstChar;

        // Примитивные парсеры.
        bind(String.class, new StringArgumentParser());
        bind(Integer.class, new IntegerArgumentParser());
        bind(int.class, new IntegerArgumentParser());
        bind(Boolean.class, new BooleanArgumentParser());
        bind(boolean.class, new BooleanArgumentParser());
        bind(Float.class, new FloatArgumentParser());
        bind(float.class, new FloatArgumentParser());
        bind(Double.class, new StringArgumentParser());
        bind(double.class, new StringArgumentParser());
    }

    /**
     * Выполнение команды, отправленной пользователем, на основе отправленного текста.
     *
     * Если команда не существует или нет такого действия, то будет выброшено исключение.
     * {@link CommandNotFoundException} CommandNotFoundException.
     *
     * Если пользователю нельзя выполнять указанную команду, то будет выброшено
     * исключение {@link CommandAccessException} CommandAccessException.
     *
     * @param sender Отправитель команды.
     * @param cmd Строка, которую отослал отправитель.
     * @return {@code true}, если команда была выполнена без ошибок.
     * @throws CommandException Выбрасывает исключение, если команды не
     * существует или нет разрешения на ее выполнение.
     */
    public CommandResult call(CommandSender sender, String cmd) throws CommandException {
        CommandContext context = validate(sender, cmd);
        CommandAction action = context.getAction();

        if(action == null) {
            throw new CommandNotFoundException();
        }

        Method method = action.getMethod();
        String[] args = context.getArgs();

        Parameter[] parameters = method.getParameters();
        int length = parameters.length;

        List<Argument<?>> arguments = new ArrayList<>();
        Object[] parsed = new Object[length];
        parsed[0] = sender;

        for(int i = 1; i < length; i++) {
            Class<?> parameterType = parameters[i].getType();
            ArgumentParser parser;

            if(Iterable.class.isAssignableFrom(parameterType) || Argument.class.isAssignableFrom(parameterType)) {
                ParameterizedType type = (ParameterizedType) parameters[i].getParameterizedType();
                parser = parsers.get(type.getActualTypeArguments()[0]);
            } else {
                parser = parsers.get(parameterType);
            }

            if(Iterable.class.isAssignableFrom(parameterType)) {
                // Подготоваливаем коллекцию.
                Collection<Object> collection = new ArrayList<>();

                for(int y = i - 1; y < args.length; y++) {
                    Argument argument = parser.parseArgument(args[y]);

                    if(Argument.class.isAssignableFrom(parameterType)) {
                        collection.add(argument);
                    } else {
                        collection.add(argument.getValue());
                    }

                    if(argument.getText() == null) {
                        argument.setText(args[y]);
                    }

                    arguments.add(argument);
                }

                parsed[i] = collection;
            } else {
                // Подготавливаем аргументы команды.
                Argument argument;

                if (args.length + 1 > i) {
                    argument = parser.parseArgument(args[i - 1]);

                    if(argument.getText() == null) {
                       argument.setText(args[i - 1]);
                    }
                } else {
                    argument = parser.parseArgument(null);
                }

                if(Argument.class.isAssignableFrom(parameterType)) {
                    parsed[i] = argument;
                } else {
                    parsed[i] = argument.getValue();
                }

                arguments.add(argument);
            }
        }

        // Выполняем команду.
        try {
            // Если возвращает void, то считаем что результат выполнения команды всегда положительный.
            Object result = method.invoke(context.getHandler().getObject(), parsed);
            boolean success = result == null || (boolean) result;
            return new CommandResult(context, arguments, success);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Автоматическое дополнение команды на основе ввода пользователя.
     *
     * @param sender Отправитель команды.
     * @param cmd Строка, которую отослал отправитель.
     * @return Возвращает список возможных вариантов автодополнения.
     */
    public Collection<String> complete(CommandSender sender, String cmd) {
        CommandContext context;

        try {
            context = validate(sender, cmd);
        } catch(CommandException e) {
            Collection<String> names = commands.keySet();

            if(removeFirstChar) {
                cmd = cmd.substring(1);
            }

            final String cmdFinal = cmd;

            if(cmd.length() > 0) {
                return names.stream().filter(name -> name.startsWith(cmdFinal)).collect(Collectors.toList());
            }

            return names;
        }

        String action = context.getActionName();
        String[] args = context.getArgs();

        if(action == null && context.getHandler().getMain() == null) {
            Collection<String> actions = context.getHandler().getActions().keySet();

            if(args.length == 1) {
                return actions.stream().filter(name -> name.startsWith(args[0])).collect(Collectors.toList());
            } else {
                return actions;
            }
        }

        String arg = "";
        int i = 1;

        if(args.length > 0) {
            i = args.length;
            arg = args[i - 1];
        }

        Parameter[] parameters = context.getAction().getMethod().getParameters();
        CommandCompleter completer;

        if(Iterable.class.isAssignableFrom(parameters[i].getType())) {
            ParameterizedType type = (ParameterizedType) parameters[i].getParameterizedType();
            completer = completers.get(type.getActualTypeArguments()[0]);
        } else {
            completer = completers.get(parameters[i].getType());
        }

        if (completer != null) {
            return completer.complete(context, arg);
        }

        return Collections.emptyList();
    }

    /**
     * Проверяет строку и конвертирует результат проверки в объект
     * {@link CommandContext} CommandContext.
     *
     * @param sender Отправитель команды.
     * @param cmd Строка, которую отослал отправитель.
     * @return Возвращает результат проверки.
     * @throws CommandException Выбрасывает исключение, если команды не
     * существует или нет разрещения на ее выполнение.
     */
    private CommandContext validate(CommandSender sender, String cmd) throws CommandException {
        checkNotNull(sender, "Please provide command sender");

        if(cmd == null || cmd.length() <= 1) {
            throw new CommandNotFoundException();
        }

        if(removeFirstChar) {
            cmd = cmd.substring(1);
        }

        String[] args = cmd.split(" ");
        CommandHandler handler = commands.get(args[0]);

        if(handler == null) {
            throw new CommandNotFoundException();
        }

        Map<String, CommandAction> actions = handler.getActions();
        CommandAction action = null;

        if(args.length > 1 && actions.containsKey(args[1])) {
            action = actions.get(args[1]);
            args = Arrays.copyOfRange(args, 2, args.length);
        } else {
            args = Arrays.copyOfRange(args, 1, args.length);

            if(handler.getMain() != null) {
                action = handler.getMain();
            }
        }

        String actionName = null;

        if(action != null) {
            Method main = handler.getMain().getMethod();
            Method method = action.getMethod();

            if(main != null) {
                actionName = main.equals(method) ? null : method.getName();
            }

            if(action.getAnnotation().permission()) {
                if (!sender.canUseCommand(handler.getName(), actionName)) {
                    throw new CommandAccessException();
                }
            }
        }

        return new CommandContext(handler.getName(), actionName, action, args, handler);
    }

    /**
     * Регистрация команды происходит по методам, которые содержат аннотацию
     * {@link Command} @Command. Количество имен для команды неограничено.
     * Обязательно должна быть хотя бы одна команда.
     *
     * @param object Объект, в котором мы ищем команды.
     * @param names Названия команд.
     */
    public void register(Object object, String... names) {
        checkNotNull(object, "Please provide command object");

        // Проверяем, что команды с таким именем еще нет.
        for(String name : names) {
            if(name == null || name.length() <= 0) {
                throw new IllegalArgumentException("Please provide valid command name");
            }

            if(CharMatcher.WHITESPACE.matchesAnyOf(name)) {
                throw new IllegalArgumentException("Please provide command name with no whitespace");
            }

            if(commands.containsKey(name)) {
                throw new IllegalArgumentException("Command with name " + name + " already exists");
            }
        }

        Map<String, CommandAction> actions = new HashMap<>();
        CommandAction main = null;

        for (Method method : object.getClass().getMethods()) {
            Command annotation = method.getAnnotation(Command.class);

            // Команда обязательно должна быть помечена аннотацией @Command.
            if(annotation == null) {
                continue;
            }

            // Команда должна обязательно возвращать void или boolean.
            if(method.getReturnType() != void.class && method.getReturnType() != boolean.class) {
                throw new IllegalArgumentException(method.getName() + " must return void or boolean");
            }

            Parameter[] parameters = method.getParameters();
            int length = parameters.length;

            // Первым параметром команды обязательно должен быть ее отправитель.
            if(length == 0 || !CommandSender.class.isAssignableFrom(parameters[0].getType())) {
                throw new IllegalArgumentException("Please provide command sender for " + method.getName());
            }

            // Проверяем, что все параметры команды будут отработаны корректно.
            for(int i = 1; i < length; i++) {
                Class<?> parameterType = parameters[i].getType();

                if(Iterable.class.isAssignableFrom(parameterType)) {
                    if (i + 1 != length) {
                        throw new IllegalArgumentException("Iterable must be last parameter in " + method.getName());
                    }

                } else {
                    if(Argument.class.isAssignableFrom(parameterType)) {
                        ParameterizedType type = (ParameterizedType) parameters[i].getParameterizedType();
                        parameterType = (Class<?>) type.getActualTypeArguments()[0];
                    }

                    if(!parsers.containsKey(parameterType)) {
                        throw new IllegalArgumentException("Please provide type parser for " + parameters[i].getType());
                    }
                }
            }

            CommandAction action = new CommandAction(method, annotation);

            if(annotation.main()) {
                main = action;
            }

            actions.put(method.getName(), action);
        }

        if(actions.isEmpty()) {
            throw new IllegalStateException("Not found any method marked with @Command");
        }

        for(String name : names) {
            commands.put(name, new CommandHandler(names[0], object, main, actions));
        }
    }

    /**
     * Удаляет все команды, связанные с выбранным классом.
     *
     * @param cls Класс, который мы удаляем из команд.
     */
    public void unregister(Class cls) {
        checkNotNull(cls, "Please provide command class");

        Iterator<CommandHandler> iterator = commands.values().iterator();

        while(iterator.hasNext()) {
            if(iterator.next().getObject().getClass().equals(cls)) {
                iterator.remove();
            }
        }
    }

    /**
     * Регистрация парсера для конвертации строки в объект.
     *
     * @param clz Класс, который мы будем конвертировать в объект.
     * @param parser Парсер, который знает как парсить класс.
     */
    public void bind(Class<?> clz, ArgumentParser parser) {
        parsers.put(clz, parser);
    }

    /**
     * Регистрация дополнителя для автодополнения команды.
     *
     * @param clz Класс, который мы будем автодополнять.
     * @param completer Дополнитель, который знает как дополнять класс.
     */
    public void bind(Class<?> clz, CommandCompleter completer) {
        completers.put(clz, completer);
    }

}

File: src/main/kotlin/ru/ensemplix/command/CommandDispatcher.kt
package ru.ensemplix.command
import ru.ensemplix.command.argument.Argument
import ru.ensemplix.command.argument.ArgumentParser
import ru.ensemplix.command.argument.ArgumentParser.*
import ru.ensemplix.command.exception.CommandAccessException
import ru.ensemplix.command.exception.CommandException
import ru.ensemplix.command.exception.CommandNotFoundException
import java.lang.reflect.ParameterizedType
import java.util.*
import java.util.stream.Collectors

/**
 * Основной класс для работы с командами.
 */
class CommandDispatcher(val removeFirstChar: Boolean = true) {

    companion object {
        /**
         * Список команд.
         */
        @JvmField
        val commands = HashMap<String, CommandHandler>()

        /**
         * Список автомачических дополнений команды.
         */
        @JvmField
        val completers = HashMap<Class<*>, CommandCompleter>()

        /**
         * Список парсеров в объекты.
         */
        @JvmField
        val parsers = HashMap<Class<*>, ArgumentParser<*>>()
    }

    init {
        // Примитивные парсеры.
        bind(String::class.java, StringArgumentParser())
        bind(Int::class.java, IntegerArgumentParser())
        bind(Integer.TYPE, IntegerArgumentParser())
        bind(Boolean::class.java, BooleanArgumentParser())
        bind(java.lang.Boolean.TYPE, BooleanArgumentParser())
        bind(Float::class.java, FloatArgumentParser())
        bind(java.lang.Float.TYPE, FloatArgumentParser())
        bind(Double::class.java, StringArgumentParser())
        bind(java.lang.Double.TYPE, StringArgumentParser())
    }

    /**
     * Выполнение команды, отправленной пользователем, на основе отправленного текста.
     *
     * Если команда не существует или нет такого действия, то будет выброшено исключение.
     * {@link CommandNotFoundException} CommandNotFoundException.
     *
     * Если пользователю нельзя выполнять указанную команду, то будет выброшено
     * исключение {@link CommandAccessException} CommandAccessException.
     *
     * @param sender Отправитель команды.
     * @param cmd Строка, которую отослал отправитель.
     * @return {@code true}, если команда была выполнена без ошибок.
     * @throws CommandException Выбрасывает исключение, если команды не
     * существует или нет разрешения на ее выполнение.
     */
    @Throws(CommandException::class)
    fun call(sender: CommandSender, cmd: String): CommandResult {
        val context = validate(sender, cmd)
        val action = context.action

        if (action == null) {
            throw CommandNotFoundException()
        }

        val method = action.method
        val args = context.args
        val parameters = method.getParameters()
        val length = parameters.size
        val arguments = ArrayList<Argument<*>>()
        val parsed = arrayOfNulls<Any>(length)
        parsed[0] = sender

        for (i in 1..length - 1) {
            val parameterType = parameters[i].getType()
            val parser: ArgumentParser<*>?

            if(Iterable::class.java.isAssignableFrom(parameterType) || Argument::class.java.isAssignableFrom(parameterType)) {
                val type = parameters[i].getParameterizedType() as ParameterizedType
                parser = parsers.get(type.getActualTypeArguments()[0])
            } else {
                parser = parsers.get(parameterType)
            }

            if(Iterable::class.java.isAssignableFrom(parameterType)) {
                // Подготоваливаем коллекцию.
                val collection = ArrayList<Any?>()

                for (y in i - 1..args.size - 1) {
                    val argument = parser!!.parseArgument(args[y])

                    if (Argument::class.java.isAssignableFrom(parameterType)) {
                        collection.add(argument)
                    } else {
                        collection.add(argument.value)
                    }

                    if(argument.text == null) {
                        argument.text = args[y]
                    }

                    arguments.add(argument)
                }
                parsed[i] = collection
            } else {
                // Подготавливаем аргументы команды.
                val argument: Argument<*>
                if(args.size + 1 > i) {
                    argument = parser!!.parseArgument(args[i - 1])
                    if (argument.text == null) {
                        argument.text = args[i - 1]
                    }
                }
                else {
                    argument = parser!!.parseArgument(null)
                }

                if (Argument::class.java.isAssignableFrom(parameterType)) {
                    parsed[i] = argument
                } else {
                    parsed[i] = argument.value
                }

                arguments.add(argument)
            }
        }
        // Выполняем команду.
        try {
            // Если возвращает void, то считаем что результат выполнения команды всегда положительный.
            val result = method.invoke(context.handler.obj, parsed)
            val success = result == null || result as Boolean
            return CommandResult(context, arguments, success)
        } catch (e:Exception) {
            throw RuntimeException(e)
        }
    }
    /**
     * Автоматическое дополнение команды на основе ввода пользователя.
     *
     * @param sender Отправитель команды.
     * @param cmd Строка, которую отослал отправитель.
     * @return Возвращает список возможных вариантов автодополнения.
     */
    fun complete(sender: CommandSender, cmd: String): Collection<String> {
        val context: CommandContext

        try {
            context = validate(sender, cmd)
        } catch (e: CommandException) {
            val names = commands.keys

            if(cmd.length > 0) {
                //return names.stream().filter({ name-> name.startsWith(cmd) }).collect(Collectors.toList())
            }

            return names
        }

        val action = context.actionName
        val args = context.args
        if(action == null && context.handler.main == null) {
            val actions = context.handler.actions.keys
            if (args.size == 1) {
                //return actions.stream().filter({ name-> name.startsWith(args[0]) }).collect(Collectors.toList())
                return actions
            } else {
                return actions
            }
        }

        var arg = ""
        var i = 1

        if(args.size > 0) {
            i = args.size
            arg = args[i - 1]
        }

        val parameters = context.action!!.method.parameters
        val parameterType = parameters[i].getType()
        val completer: CommandCompleter?

        if(Iterable::class.java.isAssignableFrom(parameterType) || Argument::class.java.isAssignableFrom(parameterType)) {
            val type = parameters[i].getParameterizedType() as ParameterizedType
            completer = completers.get(type.getActualTypeArguments()[0])
        }
        else {
            completer = completers.get(parameterType)
        }

        if (completer != null) {
            return completer.complete(context, arg)
        }

        return emptyList<String>()
    }

    /**
     * Проверяет строку и конвертирует результат проверки в объект
     * {@link CommandContext} CommandContext.
     *
     * @param sender Отправитель команды.
     * @param cmd Строка, которую отослал отправитель.
     * @return Возвращает результат проверки.
     * @throws CommandException Выбрасывает исключение, если команды не
     * существует или нет разрещения на ее выполнение.
     */
    @Throws(CommandException::class)
    private fun validate(sender: CommandSender, cmd: String?): CommandContext {
        if(cmd == null || cmd.length <= 1) {
            throw CommandNotFoundException()
        }

        var args = cmd.split((" ").toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
        val handler = commands.get(args[0])

        if(handler == null) {
            throw CommandNotFoundException()
        }

        val actions = handler.actions
        var action: CommandAction? = null

        if(args.size > 1 && actions.containsKey(args[1])) {
            action = actions.get(args[1])
            args = Arrays.copyOfRange<String>(args, 2, args.size)
        } else {
            args = Arrays.copyOfRange<String>(args, 1, args.size)

            if(handler.main != null) {
                action = handler.main
            }
        }

        var actionName: String? = null

        if (action != null) {
            val main = handler.main!!.method
            val method = action.method

            if(main != null) {
                actionName = if (main == method) null else method.getName()
            }

            if(action.annotation.permission) {
                if(!sender.canUseCommand(handler.name, actionName!!)) {
                    throw CommandAccessException()
                }
            }
        }

        return CommandContext(handler.name, actionName!!, action, args, handler)
    }
    /**
     * Регистрация команды происходит по методам, которые содержат аннотацию
     * {@link Command} @Command. Количество имен для команды неограничено.
     * Обязательно должна быть хотя бы одна команда.
     *
     * @param obj Объект, в котором мы ищем команды.
     * @param names Названия команд.
     */
    fun register(obj: Any, vararg names: String?) {
        // Проверяем, что команды с таким именем еще нет.
        for(name in names) {
            if(name == null || name.length <= 0) {
                throw IllegalArgumentException("Please provide valid command name")
            }

            //if(CharMatcher.WHITESPACE.matchesAnyOf(name)) {
            //    throw IllegalArgumentException("Please provide command name with no whitespace")
            //}

            if(commands.containsKey(name)) {
                throw IllegalArgumentException("Command with name " + name + " already exists")
            }
        }

        val actions = HashMap<String, CommandAction>()
        var main: CommandAction? = null

        for(method in obj.javaClass.methods) {
            val annotation = method.getAnnotation(Command::class.java)

            // Команда обязательно должна быть помечена аннотацией @Command.
            if(annotation == null) {
                continue
            }

            // Команда должна обязательно возвращать void или boolean.
            if(method.getReturnType() != Void.TYPE && method.getReturnType() != java.lang.Boolean.TYPE) {
                throw IllegalArgumentException(method.getName() + " must return void or boolean")
            }

            val parameters = method.getParameters()
            val length = parameters.size

            // Первым параметром команды обязательно должен быть ее отправитель.
            if(length == 0 || !CommandSender::class.java.isAssignableFrom(parameters[0].getType())) {
                throw IllegalArgumentException("Please provide command sender for " + method.getName())
            }

            // Проверяем, что все параметры команды будут отработаны корректно.
            for(i in 1..length - 1)  {
                var parameterType = parameters[i].getType()

                if(Iterable::class.java.isAssignableFrom(parameterType)) {
                    if(i + 1 != length) {
                        throw IllegalArgumentException("Iterable must be last parameter in " + method.getName())
                    }
                } else {
                    if(Argument::class.java.isAssignableFrom(parameterType)) {
                        val type = parameters[i].getParameterizedType() as ParameterizedType
                        parameterType = type.getActualTypeArguments()[0] as Class<*>
                    }

                    if(!parsers.containsKey(parameterType)) {
                        throw IllegalArgumentException("Please provide type parser for " + parameters[i].getType())
                    }
                }
            }

            val action = CommandAction(method, annotation)
            if(annotation.main) {
                main = action
            }

            actions.put(method.getName(), action)
        }

        if(actions.isEmpty()) {
            throw IllegalStateException("Not found any method marked with @Command")
        }

        for(name in names) {
            commands.put(name!!, CommandHandler(names[0]!!, obj, main!!, actions))
        }
    }

    /**
     * Удаляет все команды, связанные с выбранным классом.
     *
     * @param cls Класс, который мы удаляем из команд.
     */
    fun unregister(cls: Class<*>) {
        val iterator = commands.values.iterator()

        while(iterator.hasNext()) {
            if(iterator.next().obj.javaClass.equals(cls)) {
                iterator.remove()
            }
        }
    }

    /**
     * Регистрация парсера для конвертации строки в объект.
     *
     * @param clz Класс, который мы будем конвертировать в объект.
     * @param parser Парсер, который знает как парсить класс.
     */
    fun bind(clz: Class<*>, parser: ArgumentParser<*>) {
        parsers.put(clz, parser)
    }

    /**
     * Регистрация дополнителя для автодополнения команды.
     *
     * @param clz Класс, который мы будем автодополнять.
     * @param completer Дополнитель, который знает как дополнять класс.
     */
    fun bind(clz: Class<*>, completer: CommandCompleter) {
        completers.put(clz, completer)
    }
}
